builder/image_cache.go

```diff
@@ -0,0 +1,50 @@
+package builder
+
+import (
+	"fmt"
+
+	"github.com/docker/docker/image"
+	"github.com/docker/docker/runconfig"
+)
+
+type ImageCache struct {
+	images   map[string]*image.Image
+	children map[string]map[string]struct{} // map[parentID][childID]
+}
+
+func newImageCache(images map[string]*image.Image) *ImageCache {
+	children := make(map[string]map[string]struct{})
+	for _, img := range images {
+		if _, exists := children[img.Parent]; !exists {
+			children[img.Parent] = make(map[string]struct{})
+		}
+		children[img.Parent][img.ID] = struct{}{}
+	}
+
+	return &ImageCache{
+		images:   images,
+		children: children,
+	}
+}
+
+func (cache *ImageCache) Dispose() {
```

Comment:
I'm little confused by this method. This is not very complicated datastructure, without cycle-links, I think GC should do this job pretty well.